detect_secrets/plugins/keyword.py

```diff
@@ -52,69 +52,85 @@
 )
 # Includes ], ', " as closing
 CLOSING = r'[]\'"]{0,2}'
+AFFIX_REGEX = r'\w*'
 DENYLIST_REGEX = r'|'.join(DENYLIST)
+# Support for suffix after keyword i.e. password_secure = "value"
+DENYLIST_REGEX = r'({denylist}){suffix}'.format(
+    denylist=DENYLIST_REGEX,
+    suffix=AFFIX_REGEX,
+)
+# Support for preffix and suffix with keyword, needed for reverse comparisons
+# i.e. if ("value" == my_password_secure) {}
+DENYLIST_REGEX_WITH_PREFFIX = r'{preffix}{denylist}'.format(
+    preffix=AFFIX_REGEX,
+    denylist=DENYLIST_REGEX,
+)
 # Non-greedy match
 OPTIONAL_WHITESPACE = r'\s*'
 OPTIONAL_NON_WHITESPACE = r'[^\s]{0,50}?'
 QUOTE = r'[\'"`]'
 # Secret regex details:
-#    [^\r\n]*      -> this section match with every character except line breaks.
-#                     This allows to find secrets that starts with symbols or
-#                     alphanumeric characters.
-#    [a-zA-Z0-9]+  -> this section match only with alphanumeric characters, and at
-#                     least one is required. This allows to reduce the false positives
-#                     number.
-#    [^\r\n]*      -> this section match with every character except line breaks.
-#                     This allows to find secrets with symbols at the end.
-#    [^\r\n,\'"`]  -> this section match with the last secret character that can be
-#                     everything except line breaks, comma, backticks or quotes. This
-#                     allows to reduce the false positives number and to prevent
-#                     errors in the code snippet highlighting.
-SECRET = r'[^\r\n]*[a-zA-Z0-9]+[^\r\n]*[^\r\n,\'"`]'
+#    [^\v\g{quote_group}]*  -> this section match with every character except line breaks
+#                              and the previous quote if exists. This allows to find
+#                              secrets that starts with symbols or alphanumeric characters.
+#
+#    \w+                    -> this section match only with words (letters, numbers or _
+#                              are allowed), and at least one character is required. This
+#                              allows to reduce the false positives number.
+#
+#    [^\v\g{quote_group}]*  -> this section match with every character except line breaks
+#                              and the previous quote if exists. This allows to find secrets
+#                              with symbols at the end.
+#
+#    [^\v,\'"`]             -> this section match with the last secret character that can be
+#                              everything except line breaks, comma, backticks or quotes. This
+#                              allows to reduce the false positives number and to prevent
+#                              errors in the code snippet highlighting.
+SECRET = r'[^\v\\g{quote_group}]*\w+[^\v\\g{quote_group}]*[^\v,\'"`]'
```

Comment:
Interesting.

If I'm understanding this right, technically speaking, we can replace the "quote group" with just quotes in general (i.e. `QUOTE`), right? This assumes that secrets don't have quote like characters in them.

Reply:
Yes, we are trying to exclude the text that matches `QUOTE` but it isn't working as we expect. So we replace `\\g{quote_group}` by `\'"` in the `SECRET` regex. We think that this quotes won't be included in secrets because it is unusual in the most programming languages.